kube-proxy crash when load balancers use a different IP family

aojea opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

kube-proxy, if detects its IP family, filters out the incorrect IP version of the services to avoid issues configuring the rules.

This commit fixes a bug that crashed kube-proxy on the environment with load balancers using different IP versions that the one used by kube-proxy. It was caused by not filtering out the IPs in the LoadBalancer Status Ingress field.

Also, I've just found that the function `FilterIncorrectIPVersion` doesn't have a unit test so we should add one.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #73157

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
fix a bug in kube-proxy that caused it to crash when using load balancers with a different IP family
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea committed:
kube-proxy unit test FilterIncorrectIPVersion

Add an unit test to the kube-proxy FilterIncorrectIPVersion function

aojea commented:
/assign @thockin @danwinship

aojea commented:
/sig network

danwinship commented on pkg/proxy/util/utils_test.go:
```diff
@@ -537,3 +537,52 @@ func TestShuffleStrings(t *testing.T) {
+		},
+		{
+			ipString:        []string{"192.168.200.2", "192.1.34.23", "fd00:20::1", "2001:db9::3"},
+			expectCorrect:   []string{"192.168.200.2", "192.1.34.23"},
```
I think it would be clearer to explicitly specify `isIPv6: false` on the ipv4 tests. (Also maybe something like `wantIPv6` or `expectIPv6` or `keepIPv6`?)

otherwise looks good

aojea replied:
yeah, also will add a desc field to describe the test

aojea replied:
done

aojea commented:
/hold
I need to address Dan comments

aojea committed:
kube-proxy filter Load Balancer Status ingress

kube-proxy, if is configured with an IP family, filters out the
incorrect IP version of the services.

This commit fix a bug caused by not filtering out the IPs in the
LoadBalancer Status Ingress field.

aojea commented:
sorry for the changes, but since the fix to the bug is related to the unit test I prefer to submit only PR.
I will keep the hold until  I can test it locally, but is ready to review please take a look.

aojea commented:
/retest

aojea commented:
/hold cancel

the patch works
```
I0114 10:38:20.605719       1 server_others.go:79] IPv6 bind address (::), assume IPv6 operation
W0114 10:38:21.014615       1 server_others.go:369] Unknown proxy mode "", assuming iptables proxy
I0114 10:38:21.020153       1 node.go:135] Successfully retrieved node IP: 2001:db8:1::242:ac11:3
I0114 10:38:21.020176       1 server_others.go:150] Using iptables Proxier.
I0114 10:38:21.020371       1 server.go:580] Version: v1.18.0-alpha.1.633+8428af6fd79e1f-dirty
I0114 10:38:21.020712       1 conntrack.go:52] Setting nf_conntrack_max to 262144
I0114 10:38:21.021127       1 config.go:133] Starting endpoints config controller
I0114 10:38:21.021157       1 shared_informer.go:206] Waiting for caches to sync for endpoints config
I0114 10:38:21.021190       1 config.go:315] Starting service config controller
I0114 10:38:21.021310       1 shared_informer.go:206] Waiting for caches to sync for service config
E0114 10:38:21.023822       1 utils.go:221] 172.31.71.218 in externalIPs has incorrect IP version (service default/hello-world).
E0114 10:38:21.023876       1 utils.go:221] 2.2.2.2 in Load Balancer ingress IPs has incorrect IP version (service default/hello-world).
E0114 10:38:21.023938       1 utils.go:221] 6.6.6.6 in externalIPs has incorrect IP version (service default/testipv4).
E0114 10:38:21.024087       1 utils.go:221] 172.31.71.218 in externalIPs has incorrect IP version (service default/example-lb).
E0114 10:38:21.024105       1 utils.go:221] 5.5.5.5 in Load Balancer ingress IPs has incorrect IP version (service default/example-lb).
I0114 10:38:21.121394       1 shared_informer.go:213] Caches are synced for endpoints config 
I0114 10:38:21.121470       1 shared_informer.go:213] Caches are synced for service config 
```

aojea commented:
/priority important-soon

aojea commented:
@danwinship @thockin can you PTAL?

andrewsykim commented on pkg/proxy/util/utils_test.go:
```diff
@@ -537,3 +537,66 @@ func TestShuffleStrings(t *testing.T) {
+	for i := range testCases {
+		correct, incorrect := FilterIncorrectIPVersion(testCases[i].ipString, testCases[i].wantIPv6)
+		if !reflect.DeepEqual(testCases[i].expectCorrect, correct) {
+			t.Errorf("Test %v failed: expected %v, got %v", testCases[i].desc, testCases[i].expectCorrect, correct)
```
If you wrap this in a `t.Run` you don't have to include the test description in the error
```go
for _, testcase := range testCases {
    t.Run(testcase.desc, func(t *testing.T) {
           <rest of test>...
    })         
}
```

aojea replied:
done

andrewsykim commented on pkg/proxy/service.go:
```diff
@@ -166,6 +166,20 @@ func (sct *ServiceChangeTracker) newBaseServiceInfo(port *v1.ServicePort, servic
+		for _, ing := range service.Status.LoadBalancer.Ingress {
+			ips = append(ips, ing.IP)
+		}
+		correctIPs, incorrectIPs := utilproxy.FilterIncorrectIPVersion(ips, *sct.isIPv6Mode)
```
Should we gate this with a `if len(ips) > 0`?

aojea replied:
yeah, good catch

andrewsykim reviewed (commented):
Left some minor comments

aojea commented:
/retest

thockin commented on pkg/proxy/service_test.go:
```diff
@@ -173,8 +174,24 @@ func TestServiceToServiceMap(t *testing.T) {
-				makeServicePortName("ns1", "load-balancer", "port3", v1.ProtocolUDP): makeTestServiceInfo("172.16.55.11", 8675, "UDP", 0),
-				makeServicePortName("ns1", "load-balancer", "port4", v1.ProtocolUDP): makeTestServiceInfo("172.16.55.11", 8676, "UDP", 0),
+				makeServicePortName("ns1", "load-balancer", "port3", v1.ProtocolUDP): makeTestServiceInfo("172.16.55.11", 8675, "UDP", 0, func(info *BaseServiceInfo) {
+					info.loadBalancerStatus = v1.LoadBalancerStatus{
```
This adds a lot of vertical space to an otherwise terse test, making it hard to read.  It seems this could be:

```
info.loadBalancerStatus.Ingress = []v1.LoadBalancerIngress{{IP: "10.1.2.4"}}
```

This applies to all cases and should make this already ridiculously long list a bit shorter

thockin reviewed (commented):
Took me a minute to catch on, but seems OK.  Small request for formatting

aojea commented:
/retest
```
[sig-storage] Projected downwardAPI should provide container's memory request [NodeConformance] [Conformance] expand_more	49s
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support existing single file [LinuxOnly] expand_more	2m15s
[k8s.io] InitContainer [NodeConformance] should invoke init containers on a RestartNever pod [Conformance]
```

thockin commented:
Thanks!

/lgtm
/approve
